From e3f19ebe17d1ca02b95af5eb67198a7320319cdf Mon Sep 17 00:00:00 2001
From: Pavel Karpov <pavel.karpov@flant.com>
Date: Fri, 25 Jul 2025 15:56:21 +0300
Subject: [PATCH] add mountPermissions to shapshot

Signed-off-by: Pavel Karpov <pavel.karpov@flant.com>
---
 pkg/nfs/controllerserver.go | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/pkg/nfs/controllerserver.go b/pkg/nfs/controllerserver.go
index 3b7b21a4..9e583333 100644
--- a/pkg/nfs/controllerserver.go
+++ b/pkg/nfs/controllerserver.go
@@ -398,7 +398,7 @@ func (cs *ControllerServer) CreateSnapshot(ctx context.Context, req *csi.CreateS
 		}
 	}()
 	snapInternalVolPath := getInternalVolumePath(cs.Driver.workingMountDir, snapVol)
-	if err = os.MkdirAll(snapInternalVolPath, 0777); err != nil {
+	if err = os.MkdirAll(snapInternalVolPath, 0770); err != nil {
 		return nil, status.Errorf(codes.Internal, "failed to make subdirectory: %v", err)
 	}
 	if err := validateSnapshot(snapInternalVolPath, snapshot); err != nil {
@@ -680,6 +680,8 @@ func newNFSSnapshot(name string, params map[string]string, vol *nfsVolume) (*nfs
 			baseDir = v
 		case mountOptionsField:
 			// no op
+		case mountPermissionsField:
+			// no op
 		default:
 			return nil, status.Errorf(codes.InvalidArgument, "invalid parameter %q in snapshot storage class", k)
 		}
-- 
2.43.0

