From 75477d879cd095cd6652a6e7b33f84bdfcecfdcb Mon Sep 17 00:00:00 2001
From: Aleksandr Zimin <alexandr.zimin@flant.com>
Date: Sat, 15 Feb 2025 00:55:46 +0300
Subject: [PATCH] add volume cleanup and feature pkg

Signed-off-by: Aleksandr Zimin <alexandr.zimin@flant.com>
---
 go.mod                      |  6 ++++++
 pkg/nfs/controllerserver.go | 25 +++++++++++++++++++++++++
 2 files changed, 31 insertions(+)

diff --git a/go.mod b/go.mod
index 69f569b7..16a85539 100644
--- a/go.mod
+++ b/go.mod
@@ -26,6 +26,12 @@ require (
 	sigs.k8s.io/yaml v1.4.0
 )
 
+require (
+	github.com/deckhouse/csi-nfs/lib/go/common v0.0.0-20250213115525-4785a9da80db
+)
+
+replace github.com/deckhouse/csi-nfs/lib/go/common => ../csi-nfs/lib/go/common
+
 require (
 	github.com/NYTimes/gziphandler v1.1.1 // indirect
 	github.com/antlr/antlr4/runtime/Go/antlr/v4 v4.0.0-20230305170008-8188dc5388df // indirect
diff --git a/pkg/nfs/controllerserver.go b/pkg/nfs/controllerserver.go
index d8af15da..a5439430 100644
--- a/pkg/nfs/controllerserver.go
+++ b/pkg/nfs/controllerserver.go
@@ -258,6 +258,18 @@ func (cs *ControllerServer) DeleteVolume(ctx context.Context, req *csi.DeleteVol
 				return nil, status.Errorf(codes.Internal, "archive subdirectory(%s, %s) failed with %v", internalVolumePath, archivedInternalVolumePath, err.Error())
 			}
 		} else {
+			volumeCleanupMethod, volumeCleanupEnabled, err := getVolumeCleanupMethod(req.GetSecrets())
+			if err != nil {
+				return nil, status.Errorf(codes.Internal, "failed to get volume cleanup method: %v", err)
+			}
+
+			if volumeCleanupEnabled {
+				err = cleanupVolume(internalVolumePath, volumeCleanupMethod)
+				if err != nil {
+					return nil, status.Errorf(codes.Internal, "Volume cleanup failed with %v", err)
+				}
+			}
+
 			// delete subdirectory under base-dir
 			klog.V(2).Infof("removing subdirectory at %v", internalVolumePath)
 			if err = os.RemoveAll(internalVolumePath); err != nil {
@@ -424,6 +436,19 @@ func (cs *ControllerServer) DeleteSnapshot(ctx context.Context, req *csi.DeleteS
 
 	// delete snapshot archive
 	internalVolumePath := getInternalVolumePath(cs.Driver.workingMountDir, vol)
+
+	volumeCleanupMethod, volumeCleanupEnabled, err := getVolumeCleanupMethod(req.GetSecrets())
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to get volume cleanup method: %v", err)
+	}
+
+	if volumeCleanupEnabled {
+		err = cleanupVolume(internalVolumePath, volumeCleanupMethod)
+		if err != nil {
+			return nil, status.Errorf(codes.Internal, "Volume cleanup failed with %v", err)
+		}
+	}
+
 	klog.V(2).Infof("Removing snapshot archive at %v", internalVolumePath)
 	if err = os.RemoveAll(internalVolumePath); err != nil {
 		return nil, status.Errorf(codes.Internal, "failed to delete subdirectory: %v", err.Error())
-- 
2.39.5 (Apple Git-154)

